### PR TITLE
[APMAPI-2178] Implement ot.th and ot.rv support

### PR DESCRIPTION
**Pull request by vpellan**
**What does this PR do?**

Implement ot.th and ot.rv support in distributed tracing

**Motivation:**

Distributed tracing will work between DD-instrumented services and OTel-instrumented services, downstream OTel services will be able to forward the `th` field it to the collector and backend, which will calculate metrics correctly

**Change log entry**

Yes. Add sampling probability compatibility with OTel-instrumented services

**Additional Notes:**

**How to test the change?**

There are some system-tests that I will activate once they're merged (https://github.com/DataDog/system-tests/pull/7372)

### CODE REVIEW

File: lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        # Largest emittable threshold. `th == MAX_THRESHOLD` would mean "never sample",
+        # which is not expressible as a probability, so we clamp to one below.
+        MAX_ENCODABLE_THRESHOLD = MAX_THRESHOLD - 1
+
+        UINT64_MODULO = 1 << 64
+        UINT64_MASK = UINT64_MODULO - 1
+
+        # A well-formed threshold is 1 to 14 lowercase hex digits (a 56-bit value, trailing
+        # zero nibbles trimmed).
+        VALID_THRESHOLD = /\A[0-9a-f]{1,14}\z/
```
[the remark this answers is not in the document]
Comment (vpellan) [reply lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb]: Wrong, the OTEP do not make it an obligation to trim trailing 0, `For th keys only, trailing zeros (but not leading zeros) may be omitted.` and `th=0 -- equivalent to th=00000000000000, which is a 0% rejection threshold, which means Always Sample.`
https://github.com/open-telemetry/opentelemetry-specification/blob/v1.59.0/oteps/trace/0235-sampling-threshold-in-trace-state.md

File: lib/datadog/tracing/distributed/propagation.rb [hunk 2]
```diff
@@ -146,6 +146,9 @@ def extract(data)
                 span_id: parent_id,
                 trace_state: tracecontext_digest.trace_state,
                 trace_state_unknown_fields: tracecontext_digest.trace_state_unknown_fields,
+                trace_otel_random_value: tracecontext_digest.trace_otel_random_value,
+                trace_otel_threshold: tracecontext_digest.trace_otel_threshold,
```
[the remark this answers is not in the document]
Comment (vpellan) [reply lib/datadog/tracing/distributed/propagation.rb:150]: No, except when we do manual sampling (or ASM/AI-Guard event) we may delete the threshold. Beside these cases, if we don't originate the `ot` vendor field, we do not change it

File: lib/datadog/tracing/distributed/trace_context.rb [hunk 3]
```diff
@@ -135,40 +140,61 @@ def build_tracestate(digest)
...
-              vendors.first(TRACESTATE_MAX_LIST_MEMBERS - 1).each do |vendor|
-                break if tracestate.bytesize + vendor.bytesize + 1 > TRACESTATE_MAX_SIZE_LIMIT
+          # We are prepending our own members, so delete any existing `dd=`/`ot=`
+          # members to avoid duplicates.
+          vendors&.reject! { |v| v.start_with?("dd=", "ot=") }
 
-                tracestate << "," << vendor
-              end
-            end
+          tracestate = leading_members.join(",")
```
[the remark this answers is not in the document]
Comment (vpellan) [reply lib/datadog/tracing/distributed/trace_context.rb]: Fixed in [83df42d](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/83df42dc2f0b83c75b5ad51649f816f435340918)

File: lib/datadog/tracing/tracer.rb [hunk 4]
```diff
@@ -392,6 +394,9 @@ def build_trace(digest, auto_finish)
             tags: digest.trace_distributed_tags,
             trace_state: digest.trace_state,
             trace_state_unknown_fields: digest.trace_state_unknown_fields,
+            otel_random_value: (digest.trace_otel_random_value if propagate_sampling),
+            otel_threshold: (digest.trace_otel_threshold if propagate_sampling),
+            otel_unknown_fields: (digest.trace_otel_unknown_fields if propagate_sampling),
```
[the remark this answers is not in the document]
Comment (vpellan) [reply lib/datadog/tracing/tracer.rb]: Fix in [8983d43](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/8983d434c06d097562c15359d2a619e79fa1559a)

File: lib/datadog/tracing/sampling/otel_consistent_sampling.rb [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        end
+
+        # Parses the value of an `ot=` tracestate member (the part after `ot=`).
+        #
+        # `rv`/`th` are kept as their raw hex strings so they are forwarded byte-for-byte;
+        # other sub-keys are preserved (with a trailing `;`) for pass-through.
+        #
+        # Malformed `rv`/`th` are ignored. A random value is only meaningful alongside a
+        # threshold, so a malformed threshold also discards the random value even when it is
+        # itself well-formed.
```
Comment (mabdinur) [inline lib/datadog/tracing/sampling/otel_consistent_sampling.rb]: Nit: this comment says a malformed th also discards a well-formed rv, but the code below (and the tests) validate rv and th independently. Suggest rewording to: "Malformed rv/th are each ignored independently; a malformed th does not discard a well-formed rv."

File: spec/datadog/tracing/distributed/trace_context_spec.rb [hunk 6]
```diff
@@ -339,6 +339,65 @@
...
+        context "with 32 upstream tracestate entries" do
+          let(:options) { super().merge(trace_state: Array.new(32) { |i| "v#{i}=x" }.join(",")) }
+
+          it "reserves both leading `dd=` and `ot=`, keeping 30 upstream members" do
+            members = tracestate.split(",")
+            expect(members[0]).to eq("dd=s:1")
+            expect(members[1]).to eq("ot=rv:ef284ace7a91e1;th:e6666666666668")
+            expect(members.length).to eq(32)
+          end
+        end
```
Comment (mabdinur) [inline spec/datadog/tracing/distributed/trace_context_spec.rb:403]: Nit: no test covers the branch where dd= plus ot= together exceed the 512-byte tracestate limit and ot= gets dropped entirely (trace_context.rb:171). Worth a sibling context here.

File: spec/datadog/tracing/distributed/open_telemetry_tracestate_codec_spec.rb [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+      end
+
+      context "with a 128-bit trace id" do
+        let(:trace_id) { 0xabcdef << 64 | 0xfff972474538efff }
+
+        it "derives the random value only from the trace id's lower 64 bits" do
+          expect(resolve[0]).to eq("ef284ace7a91e1")
+        end
+      end
+    end
```
Comment (mabdinur) [inline spec/datadog/tracing/distributed/open_telemetry_tracestate_codec_spec.rb]: Nit: the limiter_dropped branch is only tested when inbound rv/th are already present. Worth a sibling context here for a purely local decision with no inbound values, expecting [nil, nil].

**Review by mabdinur — APPROVED** [review]
This looks good overall. The resolve_outbound logic matches the OTEP 235 spec, including the reconciliation for the 64-bit-to-56-bit truncation edge case. Just a few nits below, nothing blocking. We will also need to address the failing parametric tests

File: lib/datadog/tracing/sampling/otel_consistent_sampling.rb (hunk 5 above)
Comment (vpellan) [reply lib/datadog/tracing/sampling/otel_consistent_sampling.rb]: Applied in [b2908a9](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/b2908a960798f1e55bd29cfe0e79e36b07318e1e)

File: spec/datadog/tracing/distributed/trace_context_spec.rb (hunk 6 above)
Comment (vpellan) [reply spec/datadog/tracing/distributed/trace_context_spec.rb:403]: Applied in [b2908a9](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/b2908a960798f1e55bd29cfe0e79e36b07318e1e)

File: spec/datadog/tracing/distributed/open_telemetry_tracestate_codec_spec.rb (hunk 7 above)
Comment (vpellan) [reply spec/datadog/tracing/distributed/open_telemetry_tracestate_codec_spec.rb]: Thanks, applied in [b2908a9](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/b2908a960798f1e55bd29cfe0e79e36b07318e1e)

File: lib/datadog/tracing/distributed/trace_context.rb [hunk 8]
```diff
@@ -135,40 +140,67 @@ def build_tracestate(digest)
...
               # If tracestate size limit is exceeded, drop the remaining data.
-              break unless append_dd(tracestate, tag)
+              break unless append_to_vendor(dd_member, tag)
             end
           end
 
-          append_dd(tracestate, digest.trace_state_unknown_fields) if digest.trace_state_unknown_fields
+          append_to_vendor(dd_member, digest.trace_state_unknown_fields) if digest.trace_state_unknown_fields
+
+          # The OpenTelemetry consistent probability sampling member, assembled like `dd=`.
```
Comment (Strech) [inline lib/datadog/tracing/distributed/trace_context.rb]: This feels like missing programming principles here, lots of logic about the otel without any encapsulation

File: lib/datadog/tracing/distributed/trace_context.rb [hunk 9]
```diff
@@ -135,40 +140,67 @@ def build_tracestate(digest)
...
+              tracestate << "," << vendor
+            end
           end
+
+          tracestate
         end
 
         # Appends a Datadog tracestate field when it fits.
-        def append_dd(tracestate, field)
+        def append_to_vendor(tracestate, field)
```
Comment (Strech) [inline lib/datadog/tracing/distributed/trace_context.rb]: Not sure that I see any vendor here in the code

File: lib/datadog/tracing/sampling/otel_consistent_sampling.rb [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+      #   * `th` (rejection threshold): 56-bit, derived from the applied sample rate as
+      #     `round((1 - rate) * 2^56)`.
+      #
+      # This module holds the shared math so a given trace id / rate always yields
+      # identical values, matching the cross-language reference. The `ot=` member is
+      # (de)serialized alongside `dd=` in {Datadog::Tracing::Distributed::TraceContext}.
+      #
+      # @api private
+      # @see https://opentelemetry.io/docs/specs/otel/trace/tracestate-probability-sampling/
+      module OtelConsistentSampling
```
Comment (Strech) [inline lib/datadog/tracing/sampling/otel_consistent_sampling.rb]: Name is not following any existing naming in the folder

**Comment by vpellan** [comment]
**On the to_digest / Propagation - Trace Context regressions**

OTEP 235 requires deriving (or forwarding) a fresh `(rv, th)` pair on every `to_digest` call: the sampling decision is mutable trace state, so it must be snapshotted at digest-creation time rather than deferred to injection — otherwise different propagators, or repeated injections of the same digest, could observe different, possibly stale decisions.

The cost is inherent to that work: computing the OTel random value/threshold and the `ot=` tracestate member construction. Neither is incidental overhead — it's the feature doing what it's specified to do.

File: lib/datadog/tracing/sampling/otel_consistent_sampling.rb (hunk 10 above)
Comment (vpellan) [reply lib/datadog/tracing/sampling/otel_consistent_sampling.rb]: Thanks, moved it in distributed folder, where it makes more sense, and renamed it to `OpenTelemetryTracestateCodec` to follow `DatadogTagsCodec` in [66f0dbc](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/66f0dbce7d69d2ea1ce2e20e4221cbbe4ff05480)

File: lib/datadog/tracing/distributed/trace_context.rb (hunk 9 above)
Comment (vpellan) [reply lib/datadog/tracing/distributed/trace_context.rb]: I may have been confusing between members, vendors and here, tracestate. Uniformized that by renaming everything to vendors, which means this method definition is now `append_to_vendor(vendor, field)` in [66f0dbc](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/66f0dbce7d69d2ea1ce2e20e4221cbbe4ff05480)

File: lib/datadog/tracing/distributed/trace_context.rb (hunk 8 above)
Comment (vpellan) [reply lib/datadog/tracing/distributed/trace_context.rb]: Extracted it in 3 methods: `build_dd_vendor`, `build_ot_vendor` and `select_leading_vendors` in [66f0dbc](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/66f0dbce7d69d2ea1ce2e20e4221cbbe4ff05480)

File: lib/datadog/tracing/sampling/otel_consistent_sampling.rb (hunk 10 above)
Comment (Strech) [reply lib/datadog/tracing/sampling/otel_consistent_sampling.rb]: Tracestate was mentioned in the file before, now it's vendors, I have a suspicion that something is still odd

Comment (vpellan) [reply lib/datadog/tracing/sampling/otel_consistent_sampling.rb]: `tracestate` was only mentioned in comments in this file, and still is after the renaming. If you're referencing the change that was made with [this comment](https://github.com/DataDog/dd-trace-rb/pull/6111/changes/BASE..5f857203d8ce8f5b59a319ea6569b3657d87dd0a#r3713103388), it was indeed changed from `tracestate` to `vendor` because `tracestate` was incorrect: we were not appending to the tracestate, but to a specific vendor. (either `dd` or `ot`). A tracestate is made of multiple vendors.

File: lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+          # absent — downstream participants fall back to the W3C implicit random value
+          # (the trace id's 56 least-significant bits) themselves.
+          return [inbound_random_value, inbound_threshold] if inbound_threshold || inbound_random_value
+
+          # Only a trace that is making its own probability decision may derive a fresh
+          # `(rv, th)`. When a sampling priority was already assigned upstream (e.g. an
+          # older OpenTelemetry or Datadog SDK that sent no `ot` fields), DD is following
+          # that decision, not making its own, so fabricating values here would advertise a
+          # decision the origin never made: emit nothing. Same if no rate is applied.
+          return [nil, nil] if distributed_sampling_priority || !applied_rate
```
Comment (p-datadog) [inline lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb]:
```suggestion
          # that decision, not making its own: emit nothing. Same if no rate is applied.
          return [nil, nil] if distributed_sampling_priority || !applied_rate
```

File: lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+            end
+          end
+
+          threshold = nil if threshold && !VALID_THRESHOLD.match?(threshold)
+          random_value = nil if random_value && !VALID_RANDOM_VALUE.match?(random_value)
+
+          {
+            random_value: random_value,
+            threshold: threshold,
+            unknown_fields: unknown.empty? ? nil : "#{unknown.join(";")};",
```
Comment (p-datadog) [inline lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb]: Was this a conscious choice to serialize at the point of ingress rather than egress? Seems like it could have been an arbitrary LLM decision.

File: lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+        # threshold.
+        def format_threshold(threshold)
+          return "0" if threshold.zero?
+
+          trailing_nibbles = ((threshold & -threshold).bit_length - 1) / 4
+          (threshold >> (trailing_nibbles * 4)).to_s(16).rjust(14 - trailing_nibbles, "0")
+        end
+
+        private_class_method :random_value, :threshold, :reconcile_random_value,
+          :format_random_value, :format_threshold
```
Comment (p-datadog) [inline lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb]: I would put these next to each method rather than bundling at the end of file like this.

File: lib/datadog/tracing/distributed/trace_context.rb [hunk 14]
```diff
@@ -45,11 +46,12 @@ def extract(data)
 
           return unless trace_id # Could not parse traceparent
 
-          tracestate, sampling_priority, origin, ts_parent_id, tags, unknown_fields = extract_tracestate(
-            fetcher[@tracestate_key]
-          )
+          parsed = extract_tracestate(fetcher[@tracestate_key])
+          dd = parsed[:dd] || {}
+          ot = parsed[:ot] || {}
```
Comment (p-datadog) [inline lib/datadog/tracing/distributed/trace_context.rb]: suggestion (non-blocking): consider a Struct instead of indexing the hash? If fields change in the extract_tracestate this will just silently produce empty hashes.

File: lib/datadog/tracing/distributed/trace_context.rb [hunk 15]
```diff
@@ -285,28 +341,35 @@ def parse_sampled_flag(trace_flags)
...
-          dd_tracestate.slice!(0..2)
+          return {} unless vendors && !vendors.empty?
+
+          # Remove the Datadog and OpenTelemetry vendors here so they are re-emitted from the
+          # parsed values on injection rather than passed through as opaque vendors.
+          dd_vendor = pop_vendor(vendors, "dd=")
+          ot_vendor = pop_vendor(vendors, "ot=")
+
+          {
+            tracestate: vendors.join(","),
```
Comment (p-datadog) [inline lib/datadog/tracing/distributed/trace_context.rb]: This is another line that appears to serialize at ingress rather than egress.

**Review by p-datadog — APPROVED** [review]
Overall this seems fine. The usage of unstructured data seems to me like an LLM choice and it makes failures/misbehavior more difficult to diagnose, as well as the code less readable in my opinion. Up to you if you want to fix it now/later/not at all.

There are a couple of places where data is serialized at ingress. These appear to be preexisting pattern. This analysis might be helpful if you decide to investigate: https://gist.github.com/p-datadog/8722b70bdf36a00a730e62f5a2f11fac

File: lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb (hunk 12 above)
Comment (vpellan) [reply lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb]: Most likely the LLM followed the existing implementation for `dd` vendor.

File: lib/datadog/tracing/distributed/trace_context.rb (hunk 15 above)
Comment (vpellan) [reply lib/datadog/tracing/distributed/trace_context.rb]: This was already the case in the existing implementation

File: lib/datadog/tracing/tracer.rb [hunk 16]
```diff
@@ -392,7 +394,11 @@ def build_trace(digest, auto_finish)
             tags: digest.trace_distributed_tags,
             trace_state: digest.trace_state,
             trace_state_unknown_fields: digest.trace_state_unknown_fields,
+            otel_random_value: (digest.trace_otel_random_value if propagate_sampling),
+            otel_threshold: (digest.trace_otel_threshold if propagate_sampling),
+            otel_unknown_fields: digest.trace_otel_unknown_fields,
             remote_parent: digest.span_remote,
+            distributed_sampling_priority: !!sampling_priority,
             tracer: self,
```
[the remark this answers is not in the document]
Comment (vpellan) [reply lib/datadog/tracing/tracer.rb:396]: No, we should propagate unknown fields no matter what. The comment above clearly states that it only affects `ot.th` and `ot.rv`

File: lib/datadog/tracing/distributed/trace_context.rb [hunk 17]
```diff
@@ -382,16 +451,16 @@ def parse_priority_sampling(sampled, sampling_priority)
           end
         end
 
-        # We MUST NOT propagate partial members, but we SHOULD try
+        # We MUST NOT propagate partial vendors, but we SHOULD try
         # to parse as much of the tracestate as possible.
         def split_tracestate(tracestate)
           return unless tracestate
```
[the remark this answers is not in the document]
Comment (vpellan) [reply lib/datadog/tracing/distributed/trace_context.rb]: Fixed in de26c3766581cbe7ec658ce5331a98b3714b405c

File: lib/datadog/tracing/distributed/trace_context.rb (hunk 14 above)
Comment (vpellan) [reply lib/datadog/tracing/distributed/trace_context.rb]: Changed to a struct in [bcbfa62](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/bcbfa6264c5096bd20317dbe791599a4469a3829)

File: lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb (hunk 13 above)
Comment (vpellan) [reply lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb]: Changed in [0532ff7](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/0532ff7d220a8e8113970884b69b0c85f07b7bd1)

File: lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb (hunk 11 above)
Comment (vpellan) [reply lib/datadog/tracing/distributed/open_telemetry_tracestate_codec.rb]: Thanks, applied in [de26c37](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/de26c3766581cbe7ec658ce5331a98b3714b405c)

**Review by TonyCTHsu — CHANGES_REQUESTED** [review]
#### TraceContext mixes three concerns, and the data these concerns pass around needs better encapsulation.

`TraceContext` currently handles traceparent parsing (`build_traceparent`, `parse_sampled_flag`, etc.), `dd=/ot=` vendor building (`build_dd_vendor` at line 164, `build_ot_vendor` at line 189), and tracestate-container assembly (`build_tracestate` at line 130, `select_leading_vendors` at line 208, `split_tracestate` at line 449, `pop_vendor` at line 365) — all as flat private methods on one class.

This PR already moved the vendor-building logic halfway there: `build_dd_vendor`/`build_ot_vendor` are now parallel, symmetric methods, a real improvement over the old inline `dd=` string-building. But the container logic that consumes them — `build_tracestate`, `select_leading_vendors` — is still flat on `TraceContext`, mixed in with the unrelated traceparent code. Since the PR already established the vendor-builder split, I'd like to see it carried one step further: pull `build_dd_vendor`/`build_ot_vendor`/`select_leading_vendors`/`split_tracestate`/`pop_vendor` into their own container object (e.g. Tracestate), leaving `TraceContext` responsible only for traceparent parsing plus delegating to that container.

Related: `select_leading_vendors` (line 208-224) mutates its `dd_vendor`/`ot_vendor` arguments in place (`chop!`) despite reading as a pure selector — easy to miss on a second read, and worth a comment at minimum, ideally fixed as part of any restructuring.

Separately, and at a smaller scale within the same area: `otel_random_value`/`otel_threshold` are a data clump — `resolve_outbound` (open_telemetry_tracestate_codec.rb:83-127) always produces them together, `extract_otel_fields` always decides both from the same `ot=` value, and they're threaded as a pair through `TraceDigest`, `trace_operation.rb`, and `build_ot_vendor`. They should be one value object (e.g. `OtelTracestate = Struct.new(:random_value, :threshold))`, following the span_links: Array<SpanLink> precedent already on `TraceDigest`.

(`otel_unknown_fields` is fine as a separate field — independent lifecycle, matches the existing `trace_state_unknown_fields` precedent, deliberately made unconditional in "Always propagate ot unknown fields." Not part of this ask.)

#### The comments are duplicated and often placed at the wrong level.

224 lines of comments were added in this diff. The OTEP 235 explanation is written out in full independently in at least two places (open_telemetry_tracestate_codec.rb:11-24 and trace_digest.rb:86-96), and the `@see .../tracestate-probability-sampling/ link` is repeated 4 times across the diff. Worth consolidating the explanation into one place — the codec module, since that's where the actual algorithm lives — and having everything else reference it.

Separately, some comments are pinned to the wrong thing. `NON_PROBABILITY_DECISIONS` (open_telemetry_tracestate_codec.rb:52-54) carries a comment explaining why force-kept traces shouldn't carry a threshold — that's reasoning about a specific call site (`resolve_outbound`, which already has its own correct comment at lines 93-97 saying the same thing), not about what the constant itself is. A constant-level comment should describe the constant; call-site reasoning belongs at the call site, not duplicated onto the data it happens to use.

**Comment by vpellan** [comment]
By the way, I manually ran [tracing-and-appsec macrobenchmark](https://benchmarking.us1.prod.dog/benchmarks?projectId=5&page=1&ciJobDateStart=1783519633439&ciJobDateEnd=1786111633439&benchmarkGroupPipelineId=129626793&benchmarkGroupSha=b1c60b77482a24a1b07fa24bc5c50758f4282f25) and no significant regression

**Comment by TonyCTHsu** [comment]
A few naming/API observations on `TraceState` while reviewing this:

1. The `tracestate:` keyword arg in `initialize` (and the `@tracestate`/`tracestate` reader) doesn't hold the full W3C tracestate header — it's actually just the leftover vendor entries after `dd=`/`ot=` are stripped out (see `extract`, which builds it from `vendors.join(",")` after deleting the datadog/otel members). Naming it the same as the class and the raw header string is confusing. Something like `other_vendors` would better reflect what it actually is.

2. `TraceState.from_digest(digest).build` (used in `trace_context.rb` and `opentelemetry/sdk/propagator.rb`) constructs a full `TraceState` instance just to immediately call `.build` and discard the object — none of `from_digest`'s callers ever use `.datadog`/`.open_telemetry`/`.tracestate` on the result, unlike `extract`, where those readers are the whole point. Since `from_digest` is always immediately followed by `.build` at every call site, it might be worth collapsing that into a single method so the intent is clearer and there's no throwaway intermediate object.

3. Minor: `build` (on `TraceState`, `TraceState::Datadog`, and `TraceState::OpenTelemetry`) is pure serialization to a `String` with no side effects — `to_s` might read more idiomatically. One wrinkle: `TraceState#build` can return `nil` (when there's nothing to propagate), which is a bit off-contract for `to_s`'s usual "never nil" convention, while the two sub-object `build` methods already return `""` for the empty case and would fit `to_s` cleanly. Not blocking, just flagging for consideration.

**Comment by TonyCTHsu** [comment]
Thanks for making changes, the pull request is much better organized now.  Once the comments are addressed, I will approve.


One last thing to confirm/validate before merging, the marcobenchmark mentioned need to included code path that serialize/deserialized `TraceDigest` in order to proof that the overhead observed from the microbenchmark is insignificant in a macro POV.
>By the way, I manually ran [tracing-and-appsec macrobenchmark](https://benchmarking.us1.prod.dog/benchmarks?projectId=5&page=1&ciJobDateStart=1783519633439&ciJobDateEnd=1786111633439&benchmarkGroupPipelineId=129626793&benchmarkGroupSha=b1c60b77482a24a1b07fa24bc5c50758f4282f25) and no significant regression

**Comment by vpellan** [comment]
@TonyCTHsu Thanks, I applied your latest request for changes in [e48317a](https://github.com/DataDog/dd-trace-rb/pull/6111/commits/e48317a5e6c6e503165e0edbe5e2bbf47e345d90). As for the macrobenchmark, it runs on a gitlab instance but the instance is not populated with projects, so there's no request made internally when calling the currently exposed endpoint and thus no distributed traces. I changed the benchmarks in https://github.com/DataDog/benchmarking-platform/pull/302 so that it receives requests with a tracecontext, and seed the gitlab instance with a project. Once it's merged it will provide more meaningful results about this PR.

**Comment by vpellan** [comment]
I ran new macrobenchmarks that uses tracecontext (one with both extract and inject and one with only inject, meaning it will compute the th and rv) and any regression is not visible
https://benchmarking.us1.prod.dog/benchmarks?projectId=5&page=1&ciJobDateStart=1786683271017&ciJobDateEnd=1786697671017&benchmarkGroupPipelineId=130916170&benchmarkGroupSha=482821c53e27b4c8e60e9760ed60cee39251bf62

**Review by TonyCTHsu — APPROVED** [review]
